kvnemesis,kvserver: add coverage for illegal lease applied index and reproposal errors

tbg opened the pull request:
This adds testing knobs that allow probabilistically invalidating the
`MaxLeaseIndex` of proposals in the replication layer, and also allows
injecting errors into the reproposals that result from doing so.o

It then uses these knobs in kvnemesis and in a newly added "more targeted"
`kvserver` unit test. This unit test was co-developed with the kvnemesis
flavor, and various amounts of cross-pollination have ensued. The idea is that
kvnemesis produces "more randomness" and thus has a higher chance of finding
interesting behavior, while the kvserver test is more suitable to
reproducing/regression testing that behavior. For example, tracing initially
caused problems in kvnemesis, so I added tracking to the kvserver test, and
sorted out the problems this encountered (which in turn addressed the problem
in kvnemesis as well).

I am adding this testing as part of my work on
https://github.com/cockroachdb/cockroach/pull/97779. Because the test is
agnostic to how exactly reproposals are implemented, I am landing it
separately which will make it very clear that it can pass without these
changes (and once did). With #97779, the plan is to metamorphically swap
the implementations to retain coverage for both, until we are willing to
commit to the new method in #97779.

Release note: None
Epic: CRDB-25287

tbg committed:
kvserver: export PipelinedWritesEnabled

Epic: none
Release note: none

tbg committed:
kvserver: add TestProposalsWithInjectedLeaseIndexAndReproposalError

This test can shake out bugs in the reproposal pipeline by actively introducing
reproposals.

Epic: none
Release note: none

tbg committed:
kvserver: add InjectReproposalError testing knob

Epic: none
Release note: none

tbg committed:
kvserver: add LeaseFilter testing knob

This allows setting a LeaseFilter from outside the `kvserver` package.

tbg committed:
kvnemesis: fill all results on error

It hasn't bit us yet, but with upcoming code changes we will see more situations
in which a batch will hard-fail. There were a few places in
`ClosureTxnOperation` where this would leave the remaining operations in the txn
with uninitialized results. `kvnemesis` requires all results to be explicit
(i.e. success or error). Now, extending the previous but incomplete approach,
they are all set to `errOmitted`.

tbg committed:
kvnemesis: inject reproposals and reproposal failures

This introduces a flavor of single-node kvnemesis where KV mutations randomly
bounce around in the replication layer a few times before actually applying.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105272)
<!-- Reviewable:end -->

tbg commented:
No good deed goes unpunished, the new testing is bringing out some weird results:

<details>
<summary>Details</summary>

```
--- FAIL: TestKVNemesisSingleNode_ReproposalChaos (10.98s)
    test_log_scope.go:167: test logs captured to: /Users/tobias/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/3071ecfd607c2dcda710d9ffa8fdef40/logTestKVNemesisSingleNode_ReproposalChaos720228959
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:278: seed: 4043305725150643922
    kvnemesis_test.go:192: kvnemesis logging to /Users/tobias/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/3071ecfd607c2dcda710d9ffa8fdef40/kvnemesis3540977959
    kvnemesis.go:168: error applying x.Scan(ctx, tk(13779811691630448898), tk(17053492883382310021), 0) // failed to push meta={id=0d4951ff key=/Table/100/"d63f744e684afbd2" iso=ReadCommitted pri=0.04459242 epo=5 ts=1687351579.475751000,0 min=1687351579.468285000,0 seq=3} lock=true stat=PENDING rts=0,0 wto=false gul=0,0: {<nil> 0 {0x14000723090} index:0  1687351579.500642000,0}
    kvnemesis.go:180: could not scan actual latest values: injected error
        (1)
          | (opaque error wrapper)
          | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
          | reportable 0:
          |
          | github.com/cockroachdb/cockroach/pkg/kv/kvnemesis.init
          |     github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis.go:30
          | runtime.doInit
          |     GOROOT/src/runtime/proc.go:6349
          | runtime.doInit
          |     GOROOT/src/runtime/proc.go:6326
          | runtime.main
          |     GOROOT/src/runtime/proc.go:233
          | runtime.goexit
          |     GOROOT/src/runtime/asm_arm64.s:1172
        Wraps: (2) injected error
        Error types: (1) *errbase.opaqueWrapper (2) *errutil.leafError
    kvnemesis.go:188: failures(verbose): /Users/tobias/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/3071ecfd607c2dcda710d9ffa8fdef40/kvnemesis3540977959/failures
        repro steps: /Users/tobias/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/3071ecfd607c2dcda710d9ffa8fdef40/kvnemesis3540977959/repro.go
        rangefeed KVs: /Users/tobias/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/3071ecfd607c2dcda710d9ffa8fdef40/kvnemesis3540977959/kvs-rangefeed.txt
        scan KVs: 
    kvnemesis_test.go:305: 
                Error Trace:    github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:305
                                                        github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:242
                Error:          Should be zero, but was 1
                Test:           TestKVNemesisSingleNode_ReproposalChaos
                Messages:       kvnemesis detected failures
    panic.go:522: -- test log scope end --
```

</details>

I am not sure how `errInjected` even ends up in a _Scan_. Need to look at this.

tbg committed:
kvnemesis: add work-around for #105330

We're frequently getting a TransactionPushError back from `db.{Get,Scan,...}`.

Accept this as "ok" while we investigate the failure, to avoid delaying merging
the kvnemesis updates.

https://github.com/cockroachdb/cockroach/issues/105330

erikgrinaker commented on pkg/kv/kvnemesis/kvnemesis_test.go:
```diff
@@ -139,17 +209,45 @@ type kvnemesisTestCfg struct {
+	})
+}
+
+func TestKVNemesisSingleNode_ReproposalChaos(t *testing.T) {
```
I take it we don't really need any long-running nightlies for this, it's sufficient with a few runs as part of regular stress runs?

erikgrinaker commented on pkg/kv/kvnemesis/kvnemesis.go:
```diff
@@ -171,16 +174,28 @@ func RunNemesis(
+		for i := 0; ; i++ {
+			var err error
+			scanKVs, err = dbs[0].Scan(ctx, dataSpan.Key, dataSpan.EndKey, -1)
+			if errors.Is(err, errInjected) && i < 100 {
```
What's the significance of 100? Just randomly picked to terminate an infinite loop?

erikgrinaker commented on pkg/kv/kvnemesis/kvnemesis_test.go:
```diff
@@ -54,6 +56,74 @@ func testClusterArgs(tr *SeqTracker) base.TestClusterArgs {
 
+	isOurCommand := func(ba *kvpb.BatchRequest) (string, uint64, bool) {
+		key := ba.Requests[0].GetInner().Header().Key
+		n, err := fkE(string(key))
```
Should this also check that the key belongs to the data span?

tbg replied:
I thought `fkE` fully did it, it actually didn't, but adding a commit that makes it so.

erikgrinaker reviewed (approved):
Great to have coverage of this!

tbg committed:
kvnemesis: more verification in fkE

tbg commented:
Here goes, TFTR!

bors r=erikgrinaker
